File: mono/nn/unsupervised/kohonen/Node.cs
using System;

namespace nn.unsupervised.kohonen
{
	public class Node : nn.FeatureRepresentation
	{
		private int[] locations;
		private double distanceFromWinner;
		public int[] Locations
		{
			get
			{
				return locations;
			}
		}
		public double DistanceSquaredFromWinner
		{
			get
			{
				return distanceFromWinner;
			}
		}


		public Node (double[] features, int[] locations):base(features)
		{
			this.locations=locations;
		}
		public override string ToString()
		{
			System.Text.StringBuilder sb=new System.Text.StringBuilder();
			foreach (int location in locations)
			{
				sb.Append(location);
				sb.Append("\t");
			}
			//		sb.append(":");
			sb.Append(base.ToString());
			return sb.ToString();
		}
		internal override double computeSquaredFeatureDistance(double[] sample)
		{
			double result=0;
			for (int i=0;i<sample.Length;++i)
			{
				double fac=sample[i]-base.Features[i];
				result+=fac*fac;
			}
			return result;
		}
		internal void updateDistanceSquaredFromWinner(Node winner)
		{
			double result=0;
			for (int i=0;i<Locations.Length;++i)
			{
				double fac=winner.Locations[i]-locations[i];
				result+=fac*fac;
			}
			distanceFromWinner=result;
		}
		internal bool updateFeatures(double e, double deltasquared,double[] sample)
		{
			bool rv=DistanceSquaredFromWinner<deltasquared;
			if(rv==true)
			{
				double h=System.Math.Exp(-DistanceSquaredFromWinner/(2*deltasquared));
				for (int i=0;i<sample.Length;++i)
				{
					features[i]+=e*h*(sample[i]-features[i]);
				}
			}
			return rv;
		}
	}
}

File: javascript/kohonen/node.js
function Node(feat,loc) {

    this.locations = loc.slice();
    this.distanceFromWinner =0.0;
    this.features = feat.slice();


    this.computeSquaredFeatureDistance=function(sample){
    		var result=0.0;
        var i=0;
    		for (i=0;i<sample.length;++i)
    		{
    			var fac=sample[i]-this.features[i];
    			result+=fac*fac;
    		}
    		return result;
    	}

      this.updateDistanceSquaredFromWinner=function(winner)
	{
		var result=0.0;
    var i=0;
		for (i=0;i<this.locations.length;++i)
		{
			var fac=1.0*winner.locations[i]-this.locations[i];
			result+=fac*fac;
		}
		this.distanceFromWinner=result;
	}
  this.updateFeatures=function(e, deltasquared,sample)
  	{
  		var rv=this.distanceFromWinner<deltasquared;
  		if(rv==true)
  		{
  				var h=Math.exp(-this.distanceFromWinner/(2.0*deltasquared));
          var i=0;
  			for (i=0;i<sample.length;++i)
  			{
  				this.features[i]+=e*h*(sample[i]-this.features[i]);
  			}
  		}
  		return rv;
  	}
    this.toString=function(){
      var rv='';
      var i=0;
      for(i=0;i<this.locations.length;++i)
      {
        rv+=this.locations[i];
        rv+='\t';
      }
      for(i=0;i<this.features.length;++i)
      {
        rv+=this.features[i];
        rv+='\t';
      }
      return rv;
    }
}
